File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.cpp
class Solution {
public:
    int findKMostSubarrays(vector<int>& nums, int k) {
        int n = nums.size(), count = 0;
        unordered_map<int, int> mp;

        for(int i = 0, j = 0; j<n; j++) {
            mp[nums[j]]++;
            while(mp.size() > k) {
                mp[nums[i]]--;
                if(mp[nums[i]] == 0) mp.erase(nums[i]);
                i++;
            }
            count += (j - i + 1);
        }

        return count;
    }

    int subarraysWithKDistinct(vector<int>& nums, int k) {
        int ans1 = findKMostSubarrays(nums, k - 1);
        int ans2 = findKMostSubarrays(nums, k);
        return ans2 - ans1;
    }
};

File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.java
class Solution {
    public int findKmostSubarrays(int[] nums, int n, int k) {
        int count = 0;
        HashMap<Integer, Integer> mp = new HashMap<>();

        for(int i = 0, j = 0; j<n; j++) {
            mp.put(nums[j], mp.getOrDefault(nums[j], 0) + 1);
            while(mp.size() > k) {
                mp.put(nums[i], mp.get(nums[i]) - 1);
                if(mp.get(nums[i]) == 0) {
                    mp.remove(nums[i]);
                }
                i++;
            }
            count += (j - i + 1);
        }

        return count;
    }

    public int subarraysWithKDistinct(int[] nums, int k) {
        int n = nums.length;
        int ans1 = findKmostSubarrays(nums, n, k);
        int ans2 = findKmostSubarrays(nums, n, k - 1);
        return ans1 - ans2;
    }
}